File: 25-Dynamic-Programming/python/02_longest_increasing_subsequence.py
from typing import List

class Solution:
    def LIS(self, nums: List[int]) -> int:
        tails = [0] * len(nums)
        size = 0

        for num in nums:
            lo, hi = 0, size
            while lo < hi:
                mid = (lo + hi) // 2
                if tails[mid] < num:
                    lo = mid + 1
                else:
                    hi = mid
            tails[lo] = num
            if lo == size:
                size += 1

        return size

File: 25-Dynamic-Programming/java/02-LongestIncreasingSubsequence.java
class Solution {
    public int LIS(int[] nums) {
        int n = nums.length;
        int[] tails = new int[n];
        int size = 0;

        for (int num : nums) {
            int lo = 0, hi = size;
            while (lo < hi) {
                int mid = (lo + hi) / 2;
                if (tails[mid] < num) {
                    lo = mid + 1;
                } else {
                    hi = mid;
                }
            }
            tails[lo] = num;
            if (lo == size) size++;
        }

        return size;
    }
}
